opencti-platform/opencti-graphql/src/utils/stix-filtering/stix-testers.ts

```diff
@@ -0,0 +1,369 @@
+  }
+  if (stix.type === STIX_TYPE_SIGHTING) {
+    const stixValue = stix.extensions?.[STIX_EXT_OCTI].sighting_of_type;
+    const extendedStixValues: string[] = [...toValidArray(stixValue), ...getParentTypes(stixValue)];
+    return testStringFilter(filter, extendedStixValues);
+  }
+  return false;
+};
+
+/**
+ * RELATION TO TYPES
+ * - depending on stix type (relation or sighting), we might search in target_type or where_sighted_types (in extension)
+ * - we must also search in parent types
+ */
+export const testRelationToTypes = (stix: any, filter: Filter) => {
+  if (stix.type === STIX_TYPE_RELATION) {
+    const stixValue: string = stix.extensions?.[STIX_EXT_OCTI].target_type;
+    const extendedStixValues: string[] = [...toValidArray(stixValue), ...getParentTypes(stixValue)];
+    return testStringFilter(filter, extendedStixValues);
+  }
+  if (stix.type === STIX_TYPE_SIGHTING) {
+    const stixValues: string[] = stix.extensions?.[STIX_EXT_OCTI].where_sighted_types ?? [];
+    const extendedStixValues = [...stixValues, ...stixValues.map((t) => getParentTypes(t)).flat()];
+    return testStringFilter(filter, extendedStixValues);
+  }
+  return false;
+};
+
+/**
+ * REFS
+ * - uses a refs extractor using the standardId generator
+ */
+export const testRefs = (stix: any, filter: Filter) => {
+  const stixValues: string[] = stixRefsExtractor(stix, generateStandardId);
+  return testStringFilter(filter, stixValues);
+};
+
+/**
+ * INSTANCE TYPE (elementId)
+ * - elementId is id in stix extension
+ * - useSideEventMatching arg to optionally test against relations and refs of the object
+ */
+export const testInstanceType = (stix: any, filter: Filter, useSideEventMatching = false) => {
+  const stixValues: string[] = toValidArray(stix.extensions?.[STIX_EXT_OCTI]?.id);
+  if (!useSideEventMatching) {
+    // basic equality test between ids
+    return testStringFilter(filter, stixValues);
+  }
+
+  // useSideEventMatching is set ; only applies with "eq" operator
+  if (filter.operator !== 'eq') {
+    return false;
+  }
+
+  // advanced test between filter ids and the entity relations and refs
+  // we shall aggregate all candidate fields and match the filter
+  const aggregatedStixValues = [];
+  if (stix.type === STIX_TYPE_RELATION) {
+    aggregatedStixValues.push(...toValidArray(stix.target_ref)); // to
+    aggregatedStixValues.push(...toValidArray(stix.source_ref)); // from
+  }
+  if (stix.type === STIX_TYPE_SIGHTING) {
+    aggregatedStixValues.push(...(stix.where_sighted_refs ?? [])); // to
+    aggregatedStixValues.push(...toValidArray(stix.sighting_of_ref)); // from
+  }
+  // refs
+  aggregatedStixValues.push(...stixRefsExtractor(stix, generateStandardId));
+
+  return testStringFilter(filter, aggregatedStixValues);
+};
+
+/**
+ * Gives the right tester function according to the filter key.
+ * If the key is not handled, returns a function that always return false.
+ * TODO: make it dependent on the schema.
```

Comment:
The whole code in this file could be replaced by a more generic approach:
* update the schema to add the stix paths, value type and multiplicity
* extract the data thanks to the schema, according to the attribute key

And we would remove this ugly part of the code.
Now, it's far easier to say than done, so this version is MVP : it is based on the current filters available in the UI and handle each case specifically. We expect no regression. 

In a next version we could work on this complex task involving schema.

Reply:
for the record, I think it could be like the stix `converter` or the `representative` inside an entity in the schema. 

A field `stixFilter` in `AttributeDefinition` for the the individual tester logic ?

Reply:
We'll tackle this later, once a first working version is out. That's the priority.